### ISSUE TITLE
Query: Query/Tracking: entities not being tracked for complex query with nested anonymous types

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#8956]
query:

```
from e in ctx.Employees.Take(3).Select(e => new { e })
from o in ctx.Orders.Take(5).Select(o => new { o })
where e.e.EmployeeID == o.o.EmployeeID
select new { e, o }
```

only tracks one entity (employee). I would expect order to also be tracked.

test present in QueryTestBase -> Where_subquery_anon

**Comment by smitpatel** [comment]
It should not track either.
It tracks E because of ReLinq optimization causing tree to be different.

**Comment by smitpatel** [comment]
Either it should track entities inside anonymous type at any level or never track.

**Comment by maumar** [comment]
we should be tracking the entities that are projected as part of anonymous type - it is useful in case user wants to return multiple entities in one query and then perform some sort of tracking-dependent operation on them.

**Comment by maumar** [comment]
same goes for DTOs etc

**Comment by anpete** [comment]
I'm rewriting how we do tracking :smile:

**Comment by smitpatel** [comment]
There you go. :trollface:

**Comment by smitpatel** [comment]
Also consider tests in #12139

**Comment by maumar** [comment]
this is already fixed, tests enabled in b63a1eb720a6b32edd0baea9ee21028e171ab795 - closing
